### PR TITLE
ARROW-6685: [C++] Ignore trailing slashes in S3FS

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@kszucs This should fix the issue, though I haven't tested on Python.

File: cpp/src/arrow/filesystem/s3fs.cc [hunk 1]
```diff
@@ -177,17 +177,18 @@ struct S3Path {
...
   static Status FromString(const std::string& s, S3Path* out) {
-    auto first_sep = s.find_first_of(kSep);
+    const auto src = internal::RemoveTrailingSlash(s);
+    auto first_sep = src.find_first_of(kSep);
     if (first_sep == 0) {
       return Status::Invalid("Path cannot start with a separator ('", s, "')");
     }
     if (first_sep == std::string::npos) {
-      *out = {s, s, "", {}};
+      *out = {std::string(src), std::string(src), "", {}};
```
Comment (bkietz) [inline cpp/src/arrow/filesystem/s3fs.cc]: (if you prefer; this is not even a nit)
```suggestion
      *out = {src.to_string(), src.to_string(), "", {}};
```

**Review by bkietz — APPROVED** [review]
Do multiple trailing slashes ever happen? I note that RemoveTrailingSlash only removes a single slash

**Comment by pitrou** [comment]
> Do multiple trailing slashes ever happen? I note that RemoveTrailingSlash only removes a single slash

They certainly don't happen frequently. But they may happen if the user concatenates path fragments by adding slashes between them even though they are already slash-terminated.

**Comment by pitrou** [comment]
Ok, I've fixed `RemoveTrailingSlash`.

**Comment by wesm** [comment]
Travis CI build https://travis-ci.org/pitrou/arrow/builds/591604282

**Pull request closed without merging**